src/imitation/algorithms/preference_comparisons.py

```diff
@@ -906,6 +908,158 @@ def _reward_sums(self, fragment_pairs) -> Tuple[np.ndarray, np.ndarray]:
 
 
+class SynchronousCLIGatherer(PreferenceGatherer):
+    """Queries for human preferences using the command line interface."""
+
+    def __init__(
+        self,
+        video_dir: pathlib.Path,
+        custom_logger: Optional[imit_logger.HierarchicalLogger] = None,
+    ) -> None:
+        """Initialize the human preference gatherer.
+
+        Args:
+            video_dir: directory where videos of the trajectories are saved.
+            custom_logger: Where to log to; if None (default), creates a new logger.
+        """
+        super().__init__(custom_logger=custom_logger)
+        self.video_dir = video_dir
+
+    def __call__(self, fragment_pairs: Sequence[TrajectoryWithRewPair]) -> np.ndarray:
+        """Displays each pair of fragments and asks for a preference.
+
+        It iteratively requests user feedback for each pair of fragments. If in the
+        command line, it will pop out a video player for each fragment. If in a
+        notebook, it will display the videos. Either way, it will request 1 or 2 to
+        indicate which is preferred.
+
+        Args:
+            fragment_pairs: sequence of pairs of trajectory fragments
+
+        Returns:
+            A numpy array of 1 if fragment 1 is preferred and 0 otherwise, with shape
+            (b, ), where b is the length of the input
+        """
+
+        preferences = np.zeros(len(fragment_pairs), dtype=np.float32)
+        for i, (frag1, frag2) in enumerate(fragment_pairs):
+            if self._display_videos(frag1, frag2):
+                preferences[i] = 1
+        return preferences
+
+    def _display_videos(
+        self, frag1: TrajectoryWithRew, frag2: TrajectoryWithRew
+    ) -> bool:
+        """Displays the videos of the two fragments.
+
+        Args:
+            frag1: first fragment
+            frag2: second fragment
+        """
+        # display the videos
+        frag1_video_path = frag1.infos[0]["video_path"]
+        frag2_video_path = frag2.infos[0]["video_path"]
+        if self._in_ipython():
+            self._display_videos_in_notebook(frag1_video_path, frag2_video_path)
+
+            pref = input(
+                "Which video is preferred? (1 or 2, or q to quit, or r to replay): "
+            )
+            while pref not in ["1", "2", "q"]:
+                if pref == "r":
+                    self._display_videos_in_notebook(frag1_video_path, frag2_video_path)
+                pref = input("Please enter 1 or 2 or q or r: ")
+
+            if pref == "q":
+                raise KeyboardInterrupt
+            elif pref == "1":
+                return True
+            elif pref == "2":
+                return False
+
+            # should never be hit
+            raise ValueError(f"Unexpected input {pref}")
+        else:
+            print("Which video is preferred? (1 or 2, or q to quit, or r to replay):\n")
+            cap1 = cv2.VideoCapture(str(frag1_video_path))
```

Comment:
The IPython and CV sections are largely independent except for initial path extraction -- consider moving this whole `else:` branch into a helper function analogous to `self._display_videos_in_notebook`?

Reply:
Yeah I'll do that. Although the way I built it `_display_videos_in_notebook` only displays the videos, while the display and interaction code isn't cleanly separated in this code. (I could refactor to make it more modular and have the same shape but that feels like overkill.)